x/mongo/driver/mnet/connection.go

```diff
@@ -0,0 +1,127 @@
+}
+
+// WireMessageWriter represents a Connection where server operations can be
+// written to.
+type WireMessageWriter interface {
+	Write(ctx context.Context, wm []byte) error
+}
+
+// WireMessageReadWriteCloser represents a Connection where server operations
+// can read from, written to, and closed.
+type WireMessageReadWriteCloser interface {
+	WireMessageReader
+	WireMessageWriter
+	io.Closer
+}
+
+// Describer represents a Connection that can be described.
+type Describer interface {
+	Description() description.Server
+	ID() string
+	ServerConnectionID() *int64
+	DriverConnectionID() int64
+	Address() address.Address
+	Stale() bool
+}
+
+// Streamer represents a Connection that supports streaming wire protocol
+// messages using the moreToCome and exhaustAllowed flags.
+//
+// The SetStreaming and CurrentlyStreaming functions correspond to the
+// moreToCome flag on server responses. If a response has moreToCome set,
+// SetStreaming(true) will be called and CurrentlyStreaming() should return
+// true.
+//
+// CanStream corresponds to the exhaustAllowed flag. The operations layer will
+// set exhaustAllowed on outgoing wire messages to inform the server that the
+// driver supports streaming.
+type Streamer interface {
+	SetStreaming(bool)
+	CurrentlyStreaming() bool
+	SupportsStreaming() bool
+}
+
+// Compressor is an interface used to compress wire messages. If a Connection
+// supports compression it should implement this interface as well. The
+// CompressWireMessage method will be called during the execution of an
+// operation if the wire message is allowed to be compressed.
+type Compressor interface {
+	CompressWireMessage(src, dst []byte) ([]byte, error)
+}
+
+// Pinner represents a Connection that can be pinned by one or more cursors or
+// transactions. Implementations of this interface should maintain the following
+// invariants:
+//
+//  1. Each Pin* call should increment the number of references for the
+//     connection.
+//  2. Each Unpin* call should decrement the number of references for the
+//     connection.
+//  3. Calls to Close() should be ignored until all resources have unpinned the
+//     connection.
+type Pinner interface {
+	PinToCursor() error
+	PinToTransaction() error
+	UnpinFromCursor() error
+	UnpinFromTransaction() error
+}
+
+// Connection represents a connection to a MongoDB server.
+type Connection struct {
+	WireMessageReadWriteCloser
+	Describer
+	Streamer
+	Compressor
+	Pinner
+}
+
+// NewConnection creates a new Connection with the provided component. This
+// constructor returns a component that is already a Connection to avoid
+// mis-asserting the composite interfaces.
+func NewConnection(component WireMessageReadWriteCloser) *Connection {
+	if _, ok := component.(*Connection); ok {
+		return component.(*Connection)
+	}
+
+	conn := &Connection{
+		WireMessageReadWriteCloser: component,
+	}
+
+	if describer, ok := component.(Describer); ok {
+		conn.Describer = describer
+	}
```

Comment:
For all of the other interfaces (`Streamer`, `Compressor`, etc), code that calls the associated methods performs `nil` checks first. However, code that calls the `Describer` methods do not perform `nil` checks and just assumes it's available. That's a valid assumption for now, but could break unexpectedly.

There are a few options to fix that:
* Make the minimum interface be the union of `WireMessageReadWriteCloser` and `Describer`.
* Add `nil` checks to all the calls to `Describer` methods.
* Add wrapper methods for all the `Describer` method calls that handle the `nil` case. For example:
```go
func (c *Connection) Description() description.Server {
	if c.describer == nil {
		return description.Server{}
	}
	return c.describer.Description()
}
```

Reply:
I think unionizing is the best solution if we are certain that you need to describe the connection for which you're establishing. However, a `Describer` is a pretty complicated interface which makes constructing a connection more strict. For robustness, the other two solutions also seem like a good idea. What are your thoughts?

Reply:
I think 3. is probably the best change because it's easy and enables simpler connection implementations.